File: backend/app/schemas/trading.py
from __future__ import annotations

from datetime import datetime
from typing import Any, Dict, List, Literal, Optional
from pydantic import BaseModel, Field

class OptionFeeQuoteRequest(BaseModel):
    underlying_price: float = Field(..., description="Current price of underlying asset (e.g. BTC)")
    contract_size: float = Field(..., description="Size of one contract (e.g. 0.001 BTC)")
    quantity: int = Field(..., description="Number of contracts")
    premium: float = Field(..., description="Premium per contract (in quote currency)")
    order_type: Literal["maker", "taker"] = Field("taker", description="Order type: maker or taker")

class OptionFeeQuoteResponse(BaseModel):
    underlying_price: float
    contract_size: float
    quantity: int
    premium: float
    fee_rate: float
    premium_cap_rate: float
    notional: float
    notional_fee: float
    premium_value: float
    premium_cap: float
    fee: float
    applied_fee: float
    cap_applied: bool
    order_type: Literal["maker", "taker"]
    gst_rate: float
    total_fee_with_gst: float
    breakdown: Dict[str, Any]


class TradingControlRequest(BaseModel):
    action: Literal["start", "stop", "restart", "panic"]
    configuration_id: int


class TradingControlResponse(BaseModel):
    strategy_id: str | None = None
    status: Literal["starting", "running", "stopping", "stopped", "restarting", "panic", "error"]
    message: str
    stopped_sessions: int | None = None
    imported_positions: int | None = None


class SessionCleanupResponse(BaseModel):
    stopped_sessions: int
    message: str


class StrategySessionSummary(BaseModel):
    id: int
    strategy_id: str
    status: str
    activated_at: Optional[datetime]
    deactivated_at: Optional[datetime]
    duration_seconds: float | None = None
    pnl_summary: dict | None = None
    session_metadata: dict | None = None
    exit_reason: Optional[str] = None
    legs_summary: list[dict] | None = None


class StrategySessionDetail(StrategySessionSummary):
    config_snapshot: dict | None = None
    orders: list[dict]
    positions: list[dict]
    summary: dict | None = None
    monitor_snapshot: dict | None = None


class OrderLedgerRecord(BaseModel):
    order_id: str
    symbol: str
    side: str
    quantity: float
    price: float
    fill_price: float | None = None
    status: str
    created_at: datetime
    raw_response: dict | None = None


class PositionLedgerRecord(BaseModel):
    symbol: str
    side: str
    entry_price: float
    exit_price: float | None = None
    quantity: float
    realized_pnl: float | None = None
    unrealized_pnl: float | None = None
    entry_time: datetime | None = None
    exit_time: datetime | None = None
    trailing_sl_state: dict | None = None
    analytics: dict | None = None


class TradingHeartbeat(BaseModel):
    status: Literal["idle", "running", "error"]
    current_strategy_id: str | None = None
    last_heartbeat: datetime | None = None
    active_configuration_id: int | None = None
    message: str | None = None


class AnalyticsKpi(BaseModel):
    label: str
    value: float | int
    trend: float | None = None
    unit: str | None = None


class AnalyticsResponse(BaseModel):
    generated_at: datetime
    kpis: list[AnalyticsKpi]
    chart_data: dict[str, list[dict[str, float | int]]]


class AnalyticsHistoryRange(BaseModel):
    start: datetime
    end: datetime
    preset: Optional[str] = None


class AnalyticsHistoryMetrics(BaseModel):
    days_running: int = 0
    trade_count: int = 0
    win_count: int = 0
    loss_count: int = 0
    average_pnl: float = 0.0
    average_win: float = 0.0
    average_loss: float = 0.0
    win_rate: float = 0.0
    consecutive_wins: int = 0
    consecutive_losses: int = 0
    max_gain: float = 0.0
    max_loss: float = 0.0
    max_drawdown: float = 0.0
    net_pnl: float = 0.0
    pnl_before_fees: float = 0.0
    fees_total: float = 0.0
    average_fee: float = 0.0
    profitable_days: int = 0


class AnalyticsChartPoint(BaseModel):
    timestamp: datetime
    value: float
    meta: Dict[str, Any] | None = None


class AnalyticsHistogramBucket(BaseModel):
    start: float
    end: float
    count: int


class AnalyticsHistoryCharts(BaseModel):
    cumulative_pnl: List[AnalyticsChartPoint] = Field(default_factory=list)
    drawdown: List[AnalyticsChartPoint] = Field(default_factory=list)
    rolling_win_rate: List[AnalyticsChartPoint] = Field(default_factory=list)
    trades_histogram: List[AnalyticsHistogramBucket] = Field(default_factory=list)
    cumulative_gross_pnl: List[AnalyticsChartPoint] = Field(default_factory=list)
    cumulative_fees: List[AnalyticsChartPoint] = Field(default_factory=list)


class AnalyticsTimelineEntry(BaseModel):
    timestamp: datetime
    session_id: int
    order_id: Optional[str] = None
    position_id: Optional[int] = None
    symbol: Optional[str] = None
    side: Optional[str] = None
    quantity: float | None = None
    price: float | None = None
    fill_price: float | None = None
    realized_pnl: float | None = None
    unrealized_pnl: float | None = None
    metadata: Dict[str, Any] | None = None


class AnalyticsDataStatus(BaseModel):
    is_stale: bool = False
    latest_timestamp: Optional[datetime] = None
    message: Optional[str] = None


class AnalyticsHistoryResponse(BaseModel):
    generated_at: datetime
    range: AnalyticsHistoryRange
    metrics: AnalyticsHistoryMetrics
    charts: AnalyticsHistoryCharts
    timeline: List[AnalyticsTimelineEntry]
    status: AnalyticsDataStatus


class StrategyRuntimeSchedule(BaseModel):
    scheduled_entry_at: datetime | None = None
    time_to_entry_seconds: float | None = Field(
        default=None,
        description="Seconds until scheduled entry; negative if the planned time has passed.",
    )
    planned_exit_at: datetime | None = None
    time_to_exit_seconds: float | None = Field(
        default=None,
        description="Seconds until the configured exit window; negative if already elapsed.",
    )


class StrategyRuntimeTotals(BaseModel):
    realized: float = 0.0
    unrealized: float = 0.0
    total_pnl: float = 0.0
    notional: float = 0.0
    total_pnl_pct: float = 0.0
    fees: float = 0.0


class StrategyRuntimeTrailing(BaseModel):
    level: float = 0.0
    trailing_level_pct: float = 0.0
    max_profit_seen: float = 0.0
    max_profit_seen_pct: float = 0.0
    max_drawdown_seen: float = 0.0
    max_drawdown_seen_pct: float = 0.0
    enabled: bool = False


class StrategyRuntimeLimits(BaseModel):
    max_profit_pct: float = 0.0
    max_loss_pct: float = 0.0
    effective_loss_pct: float = 0.0
    trailing_enabled: bool = False
    trailing_level_pct: float = 0.0


class StrategyRuntimeSpot(BaseModel):
    entry: float | None = None
    exit: float | None = None
    last: float | None = None
    high: float | None = None
    low: float | None = None
    updated_at: datetime | None = None


class StrategyRuntimeResponse(BaseModel):
    status: Literal["idle", "waiting", "entering", "live", "cooldown"]
    mode: Optional[Literal["live", "simulation"]] = None
    active: bool
    strategy_id: str | None
    session_id: int | None
    generated_at: datetime
    schedule: StrategyRuntimeSchedule
    entry: dict | None = None
    positions: list[dict] = Field(default_factory=list)
    totals: StrategyRuntimeTotals
    limits: StrategyRuntimeLimits
    trailing: StrategyRuntimeTrailing
    spot: StrategyRuntimeSpot | None = None
    exit_reason: Optional[str] = None
    config: dict | None = None


class TradingFeeQuoteRequest(BaseModel):
    underlying_price: float = Field(..., gt=0, description="Current price of the underlying asset in USD")
    contracts: float = Field(..., gt=0, description="Number of option contracts to price")
    contract_size: float = Field(..., gt=0, description="Underlying quantity represented by a single contract")
    premium: float = Field(..., ge=0, description="Premium per unit of underlying (usually quoted in USD)")
    order_type: Literal["maker", "taker"] = Field(default="taker")
    side: Literal["buy", "sell"] = Field(default="buy")
    fee_rate: float | None = Field(
        default=None,
        ge=0,
        description="Optional override for the taker/maker fee rate (decimal form, e.g. 0.0003)",
    )
    premium_cap_rate: float | None = Field(
        default=None,
        ge=0,
        description="Optional override for the premium cap percentage (decimal form, e.g. 0.10)",
    )


class TradingFeeQuoteResponse(BaseModel):
    underlying_price: float
    contracts: float
    contract_size: float
    premium: float
    order_type: Literal["maker", "taker"]
    side: Literal["buy", "sell"]
    fee_rate: float
    premium_cap_rate: float
    notional: float
    notional_fee: float
    premium_value: float
    premium_cap: float
    fee: float
    applied_cap: Literal["notional_fee", "premium_cap"]

File: frontend/src/api/trading.ts
import client from "./client";

export interface TradingConfig {
  id: number;
  name: string;
  underlying: "BTC" | "ETH";
  delta_range_low: number;
  delta_range_high: number;
  trade_time_ist: string;
  exit_time_ist: string;
  expiry_date?: string | null;
  quantity: number;
  contract_size: number;
  max_loss_pct: number;
  max_profit_pct: number;
  trailing_sl_enabled: boolean;
  trailing_rules: Record<string, number>;
  strike_selection_mode: "delta" | "price";
  call_option_price_min: number | null;
  call_option_price_max: number | null;
  put_option_price_min: number | null;
  put_option_price_max: number | null;
  is_active: boolean;
  created_at: string;
  updated_at: string;
}

export interface TradingSessionSummary {
  id: number;
  strategy_id: string;
  status: string;
  activated_at?: string | null;
  deactivated_at?: string | null;
  duration_seconds?: number | null;
  pnl_summary?: Record<string, unknown> | null;
  session_metadata?: Record<string, unknown> | null;
  exit_reason?: string | null;
  legs_summary?: Array<Record<string, unknown>> | null;
}

export interface TradingSessionDetail extends TradingSessionSummary {
  summary?: Record<string, unknown> | null;
  monitor_snapshot?: Record<string, unknown> | null;
  orders: Array<Record<string, unknown>>;
  positions: Array<Record<string, unknown>>;
}

export interface PaginatedResponse<T> {
  items: T[];
  total: number;
  page: number;
  page_size: number;
  pages: number;
}

export interface TradingControlResult {
  strategy_id?: string;
  status: "starting" | "running" | "stopping" | "stopped" | "restarting" | "panic" | "error";
  message: string;
}

export type TradingControlAction = "start" | "stop" | "restart" | "panic";

export interface OptionFeeQuoteRequest {
  underlying_price: number;
  contract_size: number;
  quantity: number;
  premium: number;
  order_type: "maker" | "taker";
}

export interface OptionFeeQuoteResponse {
  underlying_price: number;
  contract_size: number;
  quantity: number;
  premium: number;
  fee_rate: number;
  premium_cap_rate: number;
  notional: number;
  notional_fee: number;
  premium_value: number;
  premium_cap: number;
  fee: number;
  applied_fee: number;
  cap_applied: boolean;
  order_type: "maker" | "taker";
  gst_rate: number;
  total_fee_with_gst: number;
  breakdown: Record<string, unknown>;
}

export interface AnalyticsKpi {
  label: string;
  value: number;
  trend?: number | null;
  unit?: string | null;
}

export interface AnalyticsResponse {
  generated_at: string;
  kpis: AnalyticsKpi[];
  chart_data: Record<string, Array<Record<string, number>>>;
}

export interface AnalyticsChartPoint {
  timestamp: string;
  value: number;
  meta?: Record<string, unknown> | null;
}

export interface AnalyticsHistogramBucket {
  start: number;
  end: number;
  count: number;
}

export interface AnalyticsHistoryMetrics {
  days_running: number;
  trade_count: number;
  win_count: number;
  loss_count: number;
  average_pnl: number;
  average_win: number;
  average_loss: number;
  win_rate: number;
  consecutive_wins: number;
  consecutive_losses: number;
  max_gain: number;
  max_loss: number;
  max_drawdown: number;
  net_pnl: number;
  pnl_before_fees: number;
  fees_total: number;
  average_fee: number;
  profitable_days: number;
}

export interface AnalyticsHistoryCharts {
  cumulative_pnl: AnalyticsChartPoint[];
  drawdown: AnalyticsChartPoint[];
  rolling_win_rate: AnalyticsChartPoint[];
  trades_histogram: AnalyticsHistogramBucket[];
  cumulative_gross_pnl: AnalyticsChartPoint[];
  cumulative_fees: AnalyticsChartPoint[];
}

export interface AnalyticsTimelineEntry {
  timestamp: string;
  session_id: number;
  order_id?: string | null;
  position_id?: number | null;
  symbol?: string | null;
  side?: string | null;
  quantity?: number | null;
  price?: number | null;
  fill_price?: number | null;
  realized_pnl?: number | null;
  unrealized_pnl?: number | null;
  metadata?: Record<string, unknown> | null;
}

export interface AnalyticsHistoryStatus {
  is_stale: boolean;
  latest_timestamp?: string | null;
  message?: string | null;
}

export interface AnalyticsHistoryRange {
  start: string;
  end: string;
  preset?: string | null;
}

export interface AnalyticsHistoryResponse {
  generated_at: string;
  range: AnalyticsHistoryRange;
  metrics: AnalyticsHistoryMetrics;
  charts: AnalyticsHistoryCharts;
  timeline: AnalyticsTimelineEntry[];
  status: AnalyticsHistoryStatus;
}

export interface AnalyticsHistoryParams {
  start?: string;
  end?: string;
  preset?: string;
  strategy_id?: string;
}

export interface AnalyticsExportResult {
  blob: Blob;
  filename: string;
}

export type StrategyStatus = "idle" | "waiting" | "entering" | "live" | "cooldown";

export interface RuntimeSchedule {
  scheduled_entry_at: string | null;
  time_to_entry_seconds: number | null;
  planned_exit_at: string | null;
  time_to_exit_seconds: number | null;
}

export interface RuntimeTotals {
  realized: number;
  unrealized: number;
  total_pnl: number;
  notional: number;
  total_pnl_pct: number;
  fees: number;
}

export interface RuntimeTrailing {
  level: number;
  trailing_level_pct: number;
  max_profit_seen: number;
  max_profit_seen_pct: number;
  max_drawdown_seen: number;
  max_drawdown_seen_pct: number;
  enabled: boolean;
}

export interface RuntimeLimits {
  max_profit_pct: number;
  max_loss_pct: number;
  effective_loss_pct: number;
  trailing_enabled: boolean;
  trailing_level_pct: number;
}

export interface RuntimeSpot {
  entry: number | null;
  exit: number | null;
  last: number | null;
  high: number | null;
  low: number | null;
  updated_at: string | null;
}

export interface StrategyRuntime {
  status: StrategyStatus;
  mode: "live" | "simulation" | null;
  active: boolean;
  strategy_id: string | null;
  session_id: number | null;
  generated_at: string;
  schedule: RuntimeSchedule;
  entry: Record<string, unknown> | null;
  positions: Array<Record<string, unknown>>;
  totals: RuntimeTotals;
  limits: RuntimeLimits;
  trailing: RuntimeTrailing;
  spot: RuntimeSpot | null;
  exit_reason: string | null;
  config: Record<string, unknown> | null;
}

export const fetchConfigurations = async () => {
  const response = await client.get<{ items: TradingConfig[] }>("/configs");
  return response.data.items;
};

export const createConfiguration = async (payload: Partial<TradingConfig>) => {
  const response = await client.post<TradingConfig>("/configs", payload);
  return response.data;
};

export const updateConfiguration = async (configId: number, payload: Partial<TradingConfig>) => {
  const response = await client.put<TradingConfig>(`/configs/${configId}`, payload);
  return response.data;
};

export const activateConfiguration = async (configId: number) => {
  const response = await client.post<TradingConfig>(`/configs/${configId}/activate`);
  return response.data;
};

export const deleteConfiguration = async (configId: number) => {
  await client.delete(`/configs/${configId}`);
};

export const controlTrading = async (action: TradingControlAction, configurationId: number) => {
  const response = await client.post<TradingControlResult>("/trading/control", {
    action,
    configuration_id: configurationId
  });
  return response.data;
};

export const fetchTradingSessions = async (params: { page?: number; page_size?: number } = {}) => {
  const response = await client.get<PaginatedResponse<TradingSessionSummary>>("/trading/sessions", {
    params
  });
  return response.data;
};

export const exportTradingSessionsCsv = async () => {
  const response = await client.get<Blob>("/trading/sessions/export", {
    params: { format: "csv" },
    responseType: "blob"
  });

  const dispositionRaw = typeof response.headers.get === "function"
    ? response.headers.get("content-disposition")
    : (response.headers as Record<string, string | undefined>)["content-disposition"] ??
      (response.headers as Record<string, string | undefined>)["Content-Disposition"];
  const disposition = typeof dispositionRaw === "string" ? dispositionRaw : undefined;

  const filename = parseFilenameFromDisposition(disposition) ?? "trading-sessions.csv";

  return {
    blob: response.data,
    filename
  } satisfies AnalyticsExportResult;
};

export const fetchTradingSessionDetail = async (sessionId: number) => {
  const response = await client.get<TradingSessionDetail>(`/trading/sessions/${sessionId}`);
  return response.data;
};

export const quoteOptionFees = async (payload: OptionFeeQuoteRequest) => {
  const response = await client.post<OptionFeeQuoteResponse>("/trading/fees/quote", payload);
  return response.data;
};

export const fetchAnalytics = async () => {
  const response = await client.get<AnalyticsResponse>("/analytics/dashboard");
  return response.data;
};

export const fetchAnalyticsHistory = async (params: AnalyticsHistoryParams) => {
  const response = await client.get<AnalyticsHistoryResponse>("/analytics/history", {
    params
  });
  return response.data;
};

const parseFilenameFromDisposition = (value: string | null | undefined): string | undefined => {
  if (!value) {
    return undefined;
  }

  const utf8Match = value.match(/filename\*=UTF-8''([^;]+)/i);
  if (utf8Match && utf8Match[1]) {
    try {
      return decodeURIComponent(utf8Match[1].replace(/"/g, ""));
    } catch {
      return utf8Match[1].replace(/"/g, "");
    }
  }

  const simpleMatch = value.match(/filename="?([^";]+)"?/i);
  if (simpleMatch && simpleMatch[1]) {
    return simpleMatch[1];
  }

  return undefined;
};

export const downloadAnalyticsExport = async (params: AnalyticsHistoryParams): Promise<AnalyticsExportResult> => {
  const response = await client.get<Blob>("/analytics/export", {
    params,
    responseType: "blob"
  });

  const dispositionRaw = typeof response.headers.get === "function"
    ? response.headers.get("content-disposition")
    : (response.headers as Record<string, string | undefined>)["content-disposition"] ??
      (response.headers as Record<string, string | undefined>)["Content-Disposition"];
  const disposition = typeof dispositionRaw === "string" ? dispositionRaw : undefined;

  const filename = parseFilenameFromDisposition(disposition) ?? "analytics-export.csv";

  return {
    blob: response.data,
    filename
  };
};

export const fetchRuntime = async () => {
  const response = await client.get<StrategyRuntime>("/trading/runtime");
  return response.data;
};
